### PR TITLE
add helper methods for segwit address recovery and master key retrieval

### PR DESCRIPTION
**Pull request by danielcrenna**
I added the methods I was looking for, for KeyPath recovery and master key retrieval, in case others want to do the same thing. Salt to taste. Thanks for your help!

### CODE REVIEW

**Comment by NicolasDorier** [comment]
Why are you creating `GetWalletHDKeyResponse` ?  can you just modify the current `GetWalletKeyResponse` ?

**Comment by NicolasDorier** [comment]
Also I am unsure now about the key path.

What ledger is showing you is really 0' ? I think it shows you 0'/x' where x' is the account number (so you can have multiple account in the same wallet).

Then first address of default account should be 0'/0'/0/0 first change 0'/0'/1/0

**Comment by danielcrenna** [comment]
I can modify, I made `GetWalletHDKeyResponse`  just to conform to your conventions and not do extra work to make an extended key if I don't need it / wasn't asking for it.

On my ledger each account has a root key path like `{someInteger}'/0'/x'` where x is an index into the account, as you suspected.

In my tests I can't just pass `0'/0'` to get the first account key, I need the full root key with the integer prefix. Once I have that `ExtPubKey` I can derive `0/x` to get keys for addresses.

For my purposes I don't call `GetWalletMasterPubKey`, since I don't need it, I need the key path for the account.

**Comment by NicolasDorier** [comment]
Yeah modify the existing `GetWalletKeyResponse`  instead of creating a new one.
Aside from this, it seems good to me.

I would add an optional `CancellationToken` and make `maxAttempts` nullable.
By doing this you can scan based on either, maximum attempt, or time spent.

**Comment by danielcrenna** [comment]
done

File: LedgerWallet/LedgerClient.cs [hunk 1]
```diff
@@ -60,35 +61,22 @@ public GetWalletPubKeyResponse GetWalletMasterKey()
...
-		}
-
+		
 		public KeyPath GetWalletHDKeyPathForSegwitAddress(KeyPath rootKeyPath, string segwitAddress, Network network, int startAtIndex = 0, int maxAttempts = 100)
 		{
 			return GetWalletHDKeyPathForSegwitAddressAsync(rootKeyPath, segwitAddress, network, startAtIndex, maxAttempts).GetAwaiter().GetResult();
 		}
 
-		public async Task<KeyPath> GetWalletHDKeyPathForSegwitAddressAsync(KeyPath rootKeyPath, string segwitAddress, Network network, int startAtIndex = 0, int maxAttempts = 100)
+		public async Task<KeyPath> GetWalletHDKeyPathForSegwitAddressAsync(KeyPath rootKeyPath, string segwitAddress, Network network, int startAtIndex = 0, int? maxAttempts = null, CancellationToken? cancellationToken = null)
```
Comment (NicolasDorier) [inline LedgerWallet/LedgerClient.cs]: CancellationToken cancellationToken = default(CancellationToken)

File: LedgerWallet/LedgerClient.cs [hunk 2]
```diff
@@ -102,7 +90,7 @@ public KeyPath GetWalletHDKeyPathForSegwitAddress(KeyPath rootKeyPath, string se
 				i++;
 				a++;
 
-				if (a > maxAttempts)
+				if (a > maxAttempts || (cancellationToken?.IsCancellationRequested).GetValueOrDefault())
```
Comment (NicolasDorier) [inline LedgerWallet/LedgerClient.cs]: Don't use that, just use `cancellationToken.ThrowIfCanceled();`

**Pull request merged**